pyrtl/rtllib/aes.py

```diff
@@ -79,15 +101,91 @@ def _mix_single(index):
+        mix_out = self._mix_columns(shift_out)
+        pyrtl.probe(mix_out, 'mix_row round ' + str(round))
+        key_out = pyrtl.mux(round, *key_list, default=0)
+        pyrtl.probe(key_out, 'key round ' + str(round))
+        add_round_out = self._add_round_key(add_round_in, key_out)
+        pyrtl.probe(add_round_out, 'add key round ' + str(round))
+        pyrtl.probe(cypher_text, 'plain text' + str(round))
+        with pyrtl.conditional_assignment:
+            with reset == 1:
+                round |= 0
+                key.next |= key_in
+                key_exp_in |= key_in  # to lower the number of cycles needed
+                cypher_text.next |= add_round_out
+                add_round_in |= plaintext_in
+
+            with counter == 10:  # keep everything the same
+                round |= counter
+                cypher_text.next |= cypher_text
+
+            with pyrtl.otherwise:  # running through AES
+                round |= counter + 1
+                key.next |= key
+                key_exp_in |= key
+                cypher_text.next |= add_round_out
+                with counter == 9:
+                    add_round_in |= shift_out
+                with pyrtl.otherwise:
+                    add_round_in |= mix_out
+
+        ready = (counter == 10)
+        return ready, cypher_text
+
+    def _encryption_key_gen(self, key):
+        keys = [key]
+        for enc_round in range(10):
+            key = self._key_expansion(key, enc_round)
+            keys.append(key)
+        return keys
+
+    def encryption(self, plaintext, key):
+        key_list = self._encryption_key_gen(key)
+        t = self._add_round_key(plaintext, key_list[0])
+        pyrtl.probe(t, 'add key')
+
+        for round in range(1, 11):
+            t = self._sub_bytes(t)
+            pyrtl.probe(t, 'sub' + str(round))
+            t = self._shift_rows(t)
+            pyrtl.probe(t, 'shift_row round ' + str(round))
+            if round != 10:
+                t = self._mix_columns(t)
+                pyrtl.probe(t, 'mix columns round ' + str(round))
+            # print 'after shift rows ' + t
+            t = self._add_round_key(t, key_list[round])
+        return t
+
     def decryption_statem(self, ciphertext_in, key_in, reset):
         """
-        return ready, decryption_result: ready is a one bit signal showing
-        that the answer decryption result has been calculated.
+        return ready, plain_text: ready is a one bit signal showing
+        that the decryption result (plain_text) has been calculated.
         """
         if len(key_in) != len(ciphertext_in):
             raise pyrtl.PyrtlError("AES key and ciphertext should be the same length")
 
-        cipher_text, key = (pyrtl.Register(len(ciphertext_in)) for i in range(2))
+        plain_text, key = (pyrtl.Register(len(ciphertext_in)) for i in range(2))
         key_exp_in, add_round_in = (pyrtl.WireVector(len(ciphertext_in)) for i in range(2))
```

Comment:
Umm, I don't agree with the change of cipher text to plain_text. Only at the last stage is the value in the register actually a plain_text.